### PR TITLE
store+compactor: process index cache during compaction

### PR DESCRIPTION
**Pull request by xjewer**
Add few steps during compaction:

1. Process all index cache in old blocks made by compactor until this version,
updates meta.
2. Calculate index cache during group compaction.
3. Calculate index cache during downsampling.

Store downloads index cache from object storage or process it if doesn't exist.

covers first step of #942

cc @bwplotka

### CODE REVIEW

File: pkg/block/block.go [hunk 1]
```diff
@@ -25,6 +25,8 @@ const (
 	MetaFilename = "meta.json"
 	// IndexFilename is the known index file for block index.
 	IndexFilename = "index"
+	// IndexCacheFilename is the canonical name for index cache files.
```
Comment (bwplotka) [inline pkg/block/block.go]: Nice, but let's mayb extend this a bit:
```suggestion
	// IndexCacheFilename is the canonical name for index cache file that stores essential inforrmation needed
	// to lookup postings and series' chunk later on. 
```

File: pkg/block/metadata/meta.go [hunk 2]
```diff
@@ -36,10 +38,15 @@ const (
...
 
 const (
-	// MetaVersion is a enumeration of versions supported by Thanos.
+	// MetaVersion is a enumeration of meta versions supported by Thanos.
 	MetaVersion1 = iota + 1
 )
 
+const (
+	// IndexCacheJson is a enumeration of index cache versions supported by Thanos.
+	IndexCacheJson = iota + 1
```
Comment (bwplotka) [inline pkg/block/metadata/meta.go]: `IndexCacheVersion1` rather?

Comment (bwplotka) [reply pkg/block/metadata/meta.go]: Nice point to version index cache!

File: pkg/block/metadata/meta.go [hunk 3]
```diff
@@ -56,7 +63,8 @@ type Thanos struct {
 	Downsample ThanosDownsample  `json:"downsample"`
 
 	// Source is a real upload source of the block.
-	Source SourceType `json:"source"`
+	Source SourceType   `json:"source"`
+	Cache  CacheVersion `json:"cache,omitempty"`
```
Comment (bwplotka) [inline pkg/block/metadata/meta.go]: Wait, why is that needed? I think we can drop it.

File: pkg/compact/compact.go [hunk 4]
```diff
@@ -337,6 +353,105 @@ func (c *Syncer) GarbageCollect(ctx context.Context) error {
 	return nil
 }
 
+// IndexCache creates index cache file for blocks from the bucket if blocks don't have index cache.
+func (c *Syncer) IndexCache(ctx context.Context, compactDir string) error {
```
Comment (bwplotka) [inline pkg/compact/compact.go]: This is definitely outside of syncer logic. Let's create another component, use syncer output (groups) to iterate and fill missing index caches instead. Plus we might need another one after/during compaction. Let's leave syncing as it's already quite complex. What do you think?

File: pkg/compact/compact.go [hunk 5]
```diff
@@ -101,6 +104,19 @@ func newSyncerMetrics(reg prometheus.Registerer) *syncerMetrics {
 		},
 	})
 
+	m.indexCacheTraverse = prometheus.NewCounter(prometheus.CounterOpts{
```
Comment (bwplotka) [inline pkg/compact/compact.go]: Not sure what this matric help with?

File: pkg/compact/compact.go [hunk 6]
```diff
@@ -101,6 +104,19 @@ func newSyncerMetrics(reg prometheus.Registerer) *syncerMetrics {
...
+		Name: "thanos_compact_index_cache_traverse_total",
+		Help: "Total number of index cache operations.",
+	})
+	m.indexCacheBlocks = prometheus.NewCounter(prometheus.CounterOpts{
```
Comment (bwplotka) [inline pkg/compact/compact.go]: It think this should be always equal blocks in system, right? I would say gauge of blocks with missing index cache files would be more useful? (:

File: pkg/compact/compact.go (hunk 4 above)
Comment (bwplotka) [reply pkg/compact/compact.go]: I would even be more strict:

Add index cache file ONLY on compaction. As this is weak now as we have lots of alreafy compacted blocks, we might  want to add something temporarily as here, but ONLY for blocks with compaction level > 1 no point to do that on compaction = 1 (fresh blocks). Store will already recreate those for them.

File: pkg/compact/compact.go [hunk 7]
```diff
@@ -337,6 +353,105 @@ func (c *Syncer) GarbageCollect(ctx context.Context) error {
...
+	updateMeta := func() error {
+		var newMeta *metadata.Meta
+		*newMeta = *meta
+		newMeta.Thanos.Cache = metadata.IndexCacheJson
+		if err := metadata.Write(c.logger, bdir, newMeta); err != nil {
+			return errors.Wrap(err, "write new meta")
+		}
+
+		// TODO: override meta? error on upload?
+		if err := objstore.UploadFile(ctx, c.logger, c.bkt, metaFile, metadata.MetaFilename); err != nil {
```
Comment (bwplotka) [inline pkg/compact/compact.go]: Yea, just don't do it (: Why we are adding this `meta.Thanos.Cache`? version make sense but it should part of index cache actually.

File: pkg/compact/compact.go [hunk 8]
```diff
@@ -337,6 +353,105 @@ func (c *Syncer) GarbageCollect(ctx context.Context) error {
...
+			return errors.Wrap(err, "upload index cache")
+		}
+		meta.Thanos.Cache = newMeta.Thanos.Cache
+		return nil
+	}
+
+	if ok, err := objstore.Exists(ctx, c.bkt, cache); ok {
+		return updateMeta()
+	} else if err != nil {
```
Comment (bwplotka) [inline pkg/compact/compact.go]: no need for `else`

File: pkg/compact/compact.go [hunk 9]
```diff
@@ -783,11 +898,14 @@ func (cg *Group) compact(ctx context.Context, dir string, comp tsdb.Compactor) (
...
 
 	bdir := filepath.Join(dir, compID.String())
+	index := filepath.Join(bdir, block.IndexFilename)
+	indexCache := filepath.Join(bdir, block.IndexCacheFilename)
 
 	newMeta, err := metadata.InjectThanos(cg.logger, bdir, metadata.Thanos{
 		Labels:     cg.labels.Map(),
 		Downsample: metadata.ThanosDownsample{Resolution: cg.resolution},
 		Source:     metadata.CompactorSource,
+		Cache:      metadata.IndexCacheJson,
```
Comment (bwplotka) [inline pkg/compact/compact.go]: again, this is IMO wrong.

File: pkg/compact/compact.go [hunk 10]
```diff
@@ -882,6 +1005,12 @@ func (c *BucketCompactor) Compact(ctx context.Context) error {
 			return errors.Wrap(err, "garbage")
 		}
 
+		level.Info(c.logger).Log("msg", "process index cache")
+		// Build index cache files
+		if err := c.sy.IndexCache(ctx, c.compactDir); err != nil {
```
Comment (bwplotka) [inline pkg/compact/compact.go]: This is nice, but it should be outside of syncer

Comment (bwplotka) [reply pkg/compact/compact.go]: and only for compactor lvl > 2 blocks

File: pkg/compact/downsample/streamed_block_writer.go [hunk 11]
```diff
@@ -253,10 +257,22 @@ func (w *streamedBlockWriter) writeMemPostings() error {
 	return nil
 }
 
+func (w *streamedBlockWriter) writeIndexCache() error {
+	indexFile := filepath.Join(w.blockDir, block.IndexFilename)
+	indexCacheFile := filepath.Join(w.blockDir, block.IndexCacheFilename)
+	// Build index cache file.
```
Comment (bwplotka) [inline pkg/compact/downsample/streamed_block_writer.go]: All `// Build index cache file.` comments are not needed IMO (:

File: pkg/compact/downsample/streamed_block_writer.go [hunk 12]
```diff
@@ -253,10 +257,22 @@ func (w *streamedBlockWriter) writeMemPostings() error {
...
+	}
+
+	return nil
+}
+
 // writeMetaFile writes meta file.
 func (w *streamedBlockWriter) writeMetaFile() error {
 	w.meta.Version = metadata.MetaVersion1
 	w.meta.Thanos.Source = metadata.CompactorSource
+	w.meta.Thanos.Cache = metadata.IndexCacheJson
```
Comment (bwplotka) [inline pkg/compact/downsample/streamed_block_writer.go]: kill it (:

File: pkg/store/bucket.go [hunk 13]
```diff
@@ -1052,39 +1061,50 @@ func (b *bucketBlock) loadMeta(ctx context.Context, id ulid.ULID) error {
 
 func (b *bucketBlock) loadIndexCache(ctx context.Context) (err error) {
 	cachefn := filepath.Join(b.dir, block.IndexCacheFilename)
-
-	b.indexVersion, b.symbols, b.lvals, b.postings, err = block.ReadIndexCache(b.logger, cachefn)
-	if err == nil {
+	if err := b.loadIndexCacheFromFile(ctx, cachefn); err == nil {
```
Comment (bwplotka) [inline pkg/store/bucket.go]: much better (:

File: pkg/store/bucket.go [hunk 14]
```diff
@@ -1052,39 +1061,50 @@ func (b *bucketBlock) loadMeta(ctx context.Context, id ulid.ULID) error {
...
 		return nil
 	}
 	if !os.IsNotExist(errors.Cause(err)) {
 		return errors.Wrap(err, "read index cache")
 	}
-	// No cache exists is on disk yet, build it from the downloaded index and retry.
+
+	// try to download index cache file from obj store.
```
Comment (bwplotka) [inline pkg/store/bucket.go]: Comment as sentence.
```suggestion
	// Try to download index cache file from object store.
```

File: pkg/store/bucket.go [hunk 15]
```diff
@@ -1052,39 +1061,50 @@ func (b *bucketBlock) loadMeta(ctx context.Context, id ulid.ULID) error {
...
+	if err := objstore.DownloadFile(ctx, b.logger, b.bucket, b.indexCacheFilename(), cachefn); err == nil {
+		return b.loadIndexCacheFromFile(ctx, cachefn)
+	} else if !b.bucket.IsObjNotFoundErr(err) {
```
Comment (bwplotka) [inline pkg/store/bucket.go]: not needed else

File: pkg/compact/compact.go (hunk 6 above)
Comment (xjewer) [reply pkg/compact/compact.go]: reconsider to process index cache once if no errors happened during the process. It's expected to be a temporary decision, so I guess it's ok.

File: pkg/compact/compact.go (hunk 4 above)
Comment (xjewer) [reply pkg/compact/compact.go]: >Add index cache file ONLY on compaction. As this is weak now as we have lots of alreafy compacted blocks, we might want to add something temporarily as here, but ONLY for blocks with compaction level > 1 no point to do that on compaction = 1 (fresh blocks). Store will already recreate those for them.

this is indeed a temporary solution, as you can see, it didn't process index files from sidecar.

File: cmd/thanos/compact.go [hunk 16]
```diff
@@ -84,6 +84,9 @@ func registerCompact(m map[string]setupFunc, app *kingpin.Application, name stri
 	wait := cmd.Flag("wait", "Do not exit after all compactions have been processed and wait for new work.").
 		Short('w').Bool()
 
+	processIndexCache := cmd.Flag("index.process-cache", "Process indexes' cache, upload them to obj store and update metas.").
```
Comment (bwplotka) [inline cmd/thanos/compact.go]: We made it compatible with current approach, so I would do it without any flag and always as this is recommended way (no real disadvantage)

File: cmd/thanos/compact.go [hunk 17]
```diff
@@ -186,12 +191,17 @@ func runCompact(
...
 		downsamplingDir = path.Join(dataDir, "downsample")
+		indexCacheDir   = path.Join(dataDir, "index_cache")
 	)
 
 	if err := os.RemoveAll(downsamplingDir); err != nil {
 		return errors.Wrap(err, "clean working downsample directory")
 	}
 
+	if processIndexCache {
+		sy.SetIndexCacher(compact.NewIndexCacheProcessor(logger, reg, bkt, indexCacheDir))
```
Comment (bwplotka) [inline cmd/thanos/compact.go]: again, IMO it should be enabled by default.

File: pkg/block/index.go [hunk 18]
```diff
@@ -25,19 +25,22 @@ import (
...
 	Start, End  int64
 }
 
 type indexCache struct {
-	Version     int
-	Symbols     map[uint32]string
-	LabelValues map[string][]string
-	Postings    []postingsRange
+	Version      int
+	CacheVersion int
```
Comment (bwplotka) [inline pkg/block/index.go]: Not sure if this is OK or should it be opposite: `Version` for cache version and `IndexVersion` for index version.

Comment (bwplotka) [reply pkg/block/index.go]: Also @GiedriusS if we want to add yours [JSON to bin/protobuf change](https://github.com/improbable-eng/thanos/pull/1013), should we start with that from the very cache version 1?

File: pkg/compact/compact.go [hunk 19]
```diff
@@ -140,9 +143,15 @@ func NewSyncer(logger log.Logger, reg prometheus.Registerer, bkt objstore.Bucket
 		bkt:                  bkt,
 		metrics:              newSyncerMetrics(reg),
 		blockSyncConcurrency: blockSyncConcurrency,
+		indexCacher:          NewNoopIndexCacheProcessor(),
```
Comment (bwplotka) [inline pkg/compact/compact.go]: it should be always set, no noop needed.

File: pkg/compact/index_cache.go [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type Cacher interface {
+	Cache(context.Context, map[ulid.ULID]*metadata.Meta) error
+}
+
+type noopCacheProcessor struct{}
+
+func NewNoopIndexCacheProcessor() Cacher {
```
Comment (bwplotka) [inline pkg/compact/index_cache.go]: not needed IMO

File: pkg/compact/compact.go [hunk 21]
```diff
@@ -907,6 +930,12 @@ func (c *BucketCompactor) Compact(ctx context.Context) error {
 			return errors.Wrap(err, "compaction")
 		}
 		if done {
+			// Process index cache if only all compaction done at that moment.
+			level.Info(c.logger).Log("msg", "process index cache files")
+			if err := c.sy.IndexCache(ctx); err != nil {
```
Comment (bwplotka) [inline pkg/compact/compact.go]: Again, as per https://github.com/improbable-eng/thanos/pull/986#discussion_r270219562 why we are not doing this:
* in sync? It's not related to sync at all.
* After all groups are done? It would better tied to compactor process... (: As we can do in concurrently at some point (different groups), like this:  https://github.com/improbable-eng/thanos/pull/1010

Comment (xjewer) [reply pkg/compact/compact.go]: Are u suggesting exposing blocks from Sync and pass them to CacheProcessor?

>As we can do in concurrently at some point (different groups), like this: #1010

this is one-time task, don't see any reason to make this complicated, it will change cpu/memory footprint if you don it concurrently.

File: pkg/block/index.go (hunk 18 above)
Comment (xjewer) [reply pkg/block/index.go]: I thought of it. Change could break some setups, if people preserve index cache permanently.

File: cmd/thanos/compact.go (hunk 16 above)
Comment (xjewer) [reply cmd/thanos/compact.go]: this is one-time operation, don't think the cache processor has to traverse over index files each time compactor start.

Comment (bwplotka) [reply cmd/thanos/compact.go]: Good point, but then we would like to make it explict.. when it is done? When it is safe to turn it off? Can we reword this?

Personally not sure about this now

Comment (bwplotka) [reply cmd/thanos/compact.go]: I think I like it to be a flag but we might need more explicit naming. Like `index.generate-missing-cache-files` ? essentially process cache is not very explicit.

And then it's up to a user to run it or not. Might be that user will never run it, so the oldest blocks will not have that index cache and store gw will just do it on its own.

Essentially follow up questions is: 
* How to measure completeness of this process? I guess some `bucket ls` extension that will print things without index-cache? (:

File: pkg/compact/compact.go (hunk 19 above)
Comment (bwplotka) [reply pkg/compact/compact.go]: It's weird to do composability for a plugin that is *only temporary* as we mentioned earlier. I would rather see it behind some single flag like `generateMissingIndexCacheFiles` bool

File: pkg/compact/index_cache.go (hunk 20 above)
Comment (bwplotka) [reply pkg/compact/index_cache.go]: see composability comment above

File: cmd/thanos/compact.go (hunk 17 above)
Comment (bwplotka) [reply cmd/thanos/compact.go]: ok, see my comment about composability

File: pkg/block/index.go (hunk 18 above)
Comment (bwplotka) [reply pkg/block/index.go]: Yup very good point - we want to be compatible. Thanks for being careful here :+1:

File: pkg/compact/compact.go (hunk 21 above)
Comment (bwplotka) [reply pkg/compact/compact.go]: So this comment relates to two different things, let's separate those:

> After all groups are done? It would better tied to compactor process... (: As we can do in concurrently at some point (different groups), like this: #1010

File: cmd/thanos/compact.go (hunk 16 above)
Comment (xjewer) [reply cmd/thanos/compact.go]: >index.generate-missing-cache-files

agree or maybe enable by default and the opposite flag to disable it

>How to measure completeness of this process? I guess some bucket ls extension that will print things without index-cache? (:

rely on metric `thanos_compact_index_cache_missing` ?

File: pkg/compact/compact.go (hunk 21 above)
Comment (xjewer) [reply pkg/compact/compact.go]: >Well, you don't need to do anything. After sync you already have blocks (at least meta files), so I would vote to have a function for group and run it if generateMissingIndexCacheFiles somewhere in the group loop. This will nicely go with concurrency as planned here: #1010 (actually merged) Also this relates to ONLY this temporary function that loops over group blocks and see if there is missing index cache file!

makes sense, will change

>Yes, I think we are missing long term solution here, so after compaction we would love to create the index cache file the new block before removing old blocks (: Can we do it here?

no needs to bring it here, it's already a part of compaction/downsampling proccess
https://github.com/improbable-eng/thanos/pull/986/files#diff-921a1164068d3a04217b2d85714b16adR829

https://github.com/improbable-eng/thanos/pull/986/files#diff-059ebb3a930d53936a164ef53a79d7f0R198

File: cmd/thanos/compact.go [hunk 22]
```diff
@@ -87,6 +92,9 @@ func registerCompact(m map[string]setupFunc, app *kingpin.Application, name stri
 	wait := cmd.Flag("wait", "Do not exit after all compactions have been processed and wait for new work.").
 		Short('w').Bool()
 
+	generateMissingIndexCacheFiles := cmd.Flag("index.generate-missing-cache-file", "Process indices' cache, upload them to object store and update metas.").
```
Comment (bwplotka) [inline cmd/thanos/compact.go]:
```suggestion
	generateMissingIndexCacheFiles := cmd.Flag("index.generate-missing-cache-file", "If enabled, on startup compactor runs an on-off job that scans all the blocks to find all blocks with missing index cache file. It generates those if needed and upload.").
```

File: cmd/thanos/compact.go [hunk 23]
```diff
@@ -255,6 +266,13 @@ func runCompact(
 	g.Add(func() error {
 		defer runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
 
+		// Generate index file
+		if generateMissingIndexCacheFiles {
+			if err := genMissingIndexCacheFiles(ctx, logger, bkt, indexCacheDir); err != nil {
```
Comment (bwplotka) [inline cmd/thanos/compact.go]: yes :+1:

File: cmd/thanos/compact.go [hunk 24]
```diff
@@ -255,6 +266,13 @@ func runCompact(
 	g.Add(func() error {
 		defer runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
 
+		// Generate index file
```
Comment (bwplotka) [inline cmd/thanos/compact.go]: missing trailing period ;p

File: cmd/thanos/compact.go [hunk 25]
```diff
@@ -300,3 +318,121 @@ func runCompact(
 	level.Info(logger).Log("msg", "starting compact node")
 	return nil
 }
+
+// genMissingIndexCacheFiles generates missing index cache files and uploads them to object storage.
```
Comment (bwplotka) [inline cmd/thanos/compact.go]:
```suggestion
// genMissingIndexCacheFiles scans over all blocks, generates missing index cache files and uploads them to object storage.
```

File: cmd/thanos/compact.go [hunk 26]
```diff
@@ -300,3 +318,121 @@ func runCompact(
...
+
+	defer func() {
+		if err := os.RemoveAll(dir); err != nil {
+			level.Error(logger).Log("msg", "failed to remove index cache directory", "path", dir, "err", err)
+		}
+	}()
+
+	level.Info(logger).Log("msg", "start index cache processing")
+
+	var (
```
Comment (bwplotka) [inline cmd/thanos/compact.go]:
```suggestion
	var metas []*metadata.Meta
```
is enough

File: cmd/thanos/compact.go [hunk 27]
```diff
@@ -300,3 +318,121 @@ func runCompact(
...
+		metas []*metadata.Meta
+	)
+
+	err := bkt.Iter(ctx, "", func(name string) error {
```
Comment (bwplotka) [inline cmd/thanos/compact.go]: use `if err := ... ; err != nil` idiom here maybe?

File: cmd/thanos/compact.go [hunk 28]
```diff
@@ -300,3 +318,121 @@ func runCompact(
...
+		id, ok := block.IsBlockDir(name)
+		if !ok {
+			return nil
+		}
+
+		rc, err := bkt.Get(ctx, path.Join(id.String(), block.MetaFilename))
+		if err != nil {
+			// Probably not finished block, skip it.
+			if bkt.IsObjNotFoundErr(err) {
+				level.Warn(logger).Log("msg", "meta file wasn't found", "block", id.String())
```
Comment (bwplotka) [inline cmd/thanos/compact.go]: :+1:

File: cmd/thanos/compact.go [hunk 29]
```diff
@@ -300,3 +318,121 @@ func runCompact(
...
+		}
+		defer runutil.CloseWithLogOnErr(logger, rc, "block reader")
+
+		var meta metadata.Meta
+		if err := json.NewDecoder(rc).Decode(&meta); err != nil {
+			return errors.Wrap(err, "decode meta")
+		}
+
+		// New version of compactor pushes index cache along with data block.
+		// Skip uncompacted blocks.
```
Comment (bwplotka) [inline cmd/thanos/compact.go]: :+1:

File: cmd/thanos/compact.go [hunk 30]
```diff
@@ -300,3 +318,121 @@ func runCompact(
...
+		metas = append(metas, &meta)
+
+		return nil
+	})
+	if err != nil {
+		return errors.Wrap(err, "retrieve bucket block metas")
+	}
+
+	for _, meta := range metas {
+		if err := generateIndexCacheFile(ctx, bkt, logger, dir, meta); err != nil {
```
Comment (bwplotka) [inline cmd/thanos/compact.go]: To be more verbose I would check if index cache is missing first in this loop before going to `generateIndexCacheFile`?

File: cmd/thanos/compact.go [hunk 31]
```diff
@@ -300,3 +318,121 @@ func runCompact(
...
+	defer func() {
+		if err := os.Remove(bdir); err != nil {
+			level.Error(logger).Log("msg", "failed to remove index cache directory", "path", bdir, "err", err)
+		}
+	}()
+
+	cachePath := filepath.Join(bdir, block.IndexCacheFilename)
+	cache := path.Join(meta.ULID.String(), block.IndexCacheFilename)
+
+	ok, err := objstore.Exists(ctx, bkt, cache)
```
Comment (bwplotka) [inline cmd/thanos/compact.go]: let's move it as commented above

File: pkg/compact/compact.go [hunk 32]
```diff
@@ -828,7 +831,7 @@ func (cg *Group) compact(ctx context.Context, dir string, comp tsdb.Compactor) (
 	}
 
 	// Ensure the output block is valid.
-	if err := block.VerifyIndex(cg.logger, filepath.Join(bdir, block.IndexFilename), newMeta.MinTime, newMeta.MaxTime); !cg.acceptMalformedIndex && err != nil {
+	if err := block.VerifyIndex(cg.logger, index, newMeta.MinTime, newMeta.MaxTime); err != nil {
```
Comment (bwplotka) [inline pkg/compact/compact.go]: why `!cg.acceptMalformedIndex ` was dropped?

File: pkg/objstore/objstore.go [hunk 33]
```diff
@@ -170,6 +170,23 @@ func DownloadDir(ctx context.Context, logger log.Logger, bkt BucketReader, src,
 	return nil
 }
 
+// Exists returns true, if file exists, otherwise false and nil error if presence IsObjNotFoundErr, otherwise false with
+// returning error.
+func Exists(ctx context.Context, bkt Bucket, src string) (bool, error) {
```
Comment (bwplotka) [inline pkg/objstore/objstore.go]: I don't really like this helper. Having more than two ways of doing things is usually an antipattern. It's only used in single place, can we maybe make this quick helper , local function instead? Not a blocker

**Comment by bwplotka** [comment]
+ Changelog and docs are required to tell what is happening with this flag.

File: cmd/thanos/compact.go (hunk 30 above)
Comment (xjewer) [reply cmd/thanos/compact.go]: let's keep where it is, otherwise you have to pass `IndexCacheFilename` as a parameter to `generateIndexCacheFile`.
I can rename it `generateIndexCacheFile` -> `checkIfExistsOrGenerateIndexCacheFile `

File: pkg/compact/compact.go (hunk 32 above)
Comment (xjewer) [reply pkg/compact/compact.go]: thanks, didn't notice it during resolving merge conflict.

File: pkg/objstore/objstore.go (hunk 33 above)
Comment (xjewer) [reply pkg/objstore/objstore.go]: why not?

https://github.com/improbable-eng/thanos/blob/7466e9091a600fa8029e3dea35fca5a8cc4e03c0/pkg/objstore/objstore.go#L47

you literally suggested to do the same

File: cmd/thanos/compact.go (hunk 30 above)
Comment (bwplotka) [reply cmd/thanos/compact.go]: passing another parameter is not that bad ;p

Comment (bwplotka) [reply cmd/thanos/compact.go]: but up to you

File: pkg/objstore/objstore.go (hunk 33 above)
Comment (bwplotka) [reply pkg/objstore/objstore.go]: Yea, but with explicit REMOVING statement ;p

Comment (bwplotka) [reply pkg/objstore/objstore.go]: But you are right, let's leave it and  remove EXISTS from interface in another PRs, thanks for pointing this (:

File: CHANGELOG.md [hunk 34]
```diff
@@ -35,6 +36,7 @@ New tracing span:
 - [#970](https://github.com/improbable-eng/thanos/pull/970) Added `PartialResponseStrategy` field for `RuleGroups` for `Ruler`.
 - [#1016](https://github.com/improbable-eng/thanos/pull/1016) Added option for another DNS resolver (miekg/dns client). 
 This to have SRV resolution working on [Golang 1.11+ with KubeDNS below v1.14](https://github.com/golang/go/issues/27546)
+- [#986](https://github.com/improbable-eng/thanos/pull/986) Store index cache files in object storage, reduces store start-up time by skipping the generating the index cache for all blocks and only do this for recently created uncompacted blocks.
```
Comment (bwplotka) [inline CHANGELOG.md]: Could be reworded, but we can do it later on cut 0.4.0 :+1:

File: cmd/thanos/compact.go (hunk 30 above)
Comment (bwplotka) [reply cmd/thanos/compact.go]: not addressed I think

**Comment by GiedriusS** [comment]
Wasn't fast enough with a full review before you merged it but it looks really nice 👍 awesome, thank you a lot!

**Comment by bwplotka** [comment]
Feel free to mark your feedback somewhere! We can include it in later PR
